src/PathMatcher/CompositeMatcher.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Files\PathMatcher;
+
+use InvalidArgumentException;
+use function get_class;
+use function is_object;
+
+/**
+ * Composite matcher allows combining several matchers.
+ */
+final class CompositeMatcher implements PathMatcherInterface
+{
+    private bool $matchAny;
+
+    /**
+     * @var array PathMatcherInterface[]
+     */
+    private array $matchers;
+
+    private function __construct(bool $matchAny, array $matchers)
+    {
+        foreach ($matchers as $matcher) {
+            if (!$matcher instanceof PathMatcherInterface) {
+                $type = is_object($matcher) ? get_class($matcher) : gettype($matcher);
+                $message = sprintf(
+                    'Matchers should contain instances of \Yiisoft\Files\PathMatcher\PathMatcherInterface, %s given.',
+                    $type
+                );
+                throw new InvalidArgumentException($message);
+            }
+        }
+        $this->matchers = $matchers;
+        $this->matchAny = $matchAny;
+    }
```

Comment:
```suggestion
    private function __construct(bool $matchAny, PathMatcherInterface ...$matchers)
    {
        $this->matchers = $matchers;
        $this->matchAny = $matchAny;
    }
```
I think typed variadic is more convenient way to check list items type